Adds test that checks for duplicate APCs in an area.

atlantiscze opened the pull request:
- Should prevent this from happening in the future.
- This commit will most likely fail, as there are currently duplicate APCs present. There may also be bugs in the code, i'm shamelessly using Travis to test it for me.

atlantiscze committed:
Adds test that checks for duplicate APCs in an area.
- Should prevent this from happening in the future.
- This commit will most likely fail, as there are currently duplicate APCs present. There may also be bugs in the code, i'm shamelessly using Travis to test it for me.

PsiOmegaDelta commented on code/unit_tests/power_tests.dm:
```diff
@@ -46,4 +46,17 @@ datum/unit_test/roundstart_cable_connectivity/start_test()
+			if(!found_apc)
+				found_apc = APC
+				continue
+			fail("Found duplicated APCs in area: [A.name], first: x[found_apc.x] y[found_apc.y] z[found_apc.z]  second: y[APC.x] y[APC.y] z[APC.z]")
```
log_info_line (found_apc)
log_info_line(APC)

atlantiscze committed:
Fixes torch map, updates test
- Updates test to not runtime at the end.
- Updates torch maps to remove all duplicate APCs found by previous run of the test.
- Updates some other Z levels by running them through JMerge, which should prevent automatic merges from failing (until someone breaks them yet again). No player-facing changes here.
- Fixes #17549 and prevents it from re-occuring in the future.

PsiOmegaDelta commented on code/unit_tests/power_tests.dm:
```diff
@@ -38,12 +38,27 @@ datum/unit_test/roundstart_cable_connectivity/start_test()
+			if(!found_apc)
+				found_apc = APC
+				continue
+			fail("Found duplicated APCs in area: [A.name], first: \[[found_apc.x],[found_apc.y],[found_apc.z]\] second: \[[APC.x],[APC.y],[APC.z]\]")
```
log_info_line (found_apc) and log_info_line(APC) will probably be useful here.

atlantiscze committed:
Fixes remaining occurences of duplicated APCs, updates test a bit more

PsiOmegaDelta commented on code/unit_tests/power_tests.dm:
```diff
@@ -38,12 +38,29 @@ datum/unit_test/roundstart_cable_connectivity/start_test()
+				continue
+			fail("Duplicated APCs in area: [A.name]. Details: ")
+			log_info_line(found_apc)
+			log_info_line(APC)
```
Ah, no. This proc produces strings that are designed to be used with debug output.
I.e. ```fail("Duplicated APCs in area: [A.name]. Details: [log_info_line(found_apc)]") - [log_info_line(APC)]```, or how you now prefer to format it.

atlantiscze replied:
That's what i get for not testing it locally.

atlantiscze committed:
Fixes unit test output

atlantiscze commented:
Should be resolved.

mustafakalash commented on code/unit_tests/power_tests.dm:
```diff
@@ -38,12 +38,28 @@ datum/unit_test/roundstart_cable_connectivity/start_test()
+			if(!found_apc)
+				found_apc = APC
+				continue
+			fail("Duplicated APCs in area: [A.name]. #1: [log_info_line(found_apc)]  #2: [log_info_line(APC)]")
```
Return here, just in case we somehow get three or more APCs in one area, we don't want to call `fail()` more than once.

atlantiscze commented:
Adding a return here would mean that the unit test will only detect a 
single pair of duplicate APCs. If there's two or more pairs of 
duplicated APCs, it won't be detected.

Adding a continue instead would be plausible, though i would personally 
prefer if it reported all duplicate pairs - in the unlikely case that 
someone maps in three or more APCs in an area.

Also, this is ready for merge, ignore the DNM tag in title. I am at work 
right now, and i can't change it via email reply.

mustafakalash commented:
The problem is that calling `fail()` again will report that more tests failed than will actually have. You should adjust the check to print all excess APCs without calling `fail()` more than once.

PsiOmegaDelta commented:
What you can do is group APCs by area, and then use a helper method to report any with excess issues.
https://github.com/Baystation12/Baystation12/blob/dev/code/unit_tests/unique_tests.dm have a couple of examples, see uses of ```group_by()``` and ```number_of_issues()```.

atlantiscze commented:
I'll fix this tommorow, i've been too busy today, unfortunately.

atlantiscze committed:
Changes unit test output.

atlantiscze commented:
Just for the record, thing mentioned in @mustafakalash 's review has been addressed.
